airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/load/task/DestinationTaskScopeProvider.kt

```diff
@@ -64,31 +71,42 @@ class DestinationTaskScopeProvider(config: DestinationConfiguration) :
                 .asCoroutineDispatcher()
         )
 
-    override suspend fun launch(task: ScopedTask) {
-        when (task) {
-            is InternalTask -> internalScope.scope.launch { execute(task) }
-            is ImplementorTask -> implementorScope.scope.launch { execute(task) }
+    override suspend fun launch(task: WrappedTask<ScopedTask>) {
+        when (task.innerTask) {
+            is InternalScope -> internalScope.scope.launch { execute(task, "internal") }
+            is ImplementorScope -> implementorScope.scope.launch { execute(task, "implementor") }
+            is ShutdownScope -> implementorScope.scope.launch { execute(task, "shutdown") }
         }
     }
 
-    private suspend fun execute(task: ScopedTask) {
-        log.info { "Launching task $task" }
+    private suspend fun execute(task: WrappedTask<ScopedTask>, scope: String) {
+        log.info { "Launching task $task in scope $scope" }
         val elapsed = measureTimeMillis { task.execute() }
         log.info { "Task $task completed in $elapsed ms" }
     }
 
-    override suspend fun close() = supervisorScope {
+    override suspend fun close() {
         log.info { "Closing task scopes" }
-        internalScope.job.cancel()
         // Under normal operation, all tasks should be complete
         // (except things like force flush, which loop). So
         // - it's safe to force cancel the internal tasks
-        // - implementor scope should join immediately unless we're
-        //   failing, in which case we want to give them a chance to
-        //   fail gracefully
+        // - implementor scope should join immediately
+        internalScope.job.cancel()
+        log.info { "Cancelled internal tasks, waiting for implementor tasks to complete" }
+        implementorScope.job.join()
+        log.info { "Implementor tasks completed" }
+    }
+
+    override suspend fun kill() {
+        log.info { "Killing task scopes" }
+        internalScope.job.cancel()
+        // Give the implementor tasks a chance to fail gracefully
         withTimeoutOrNull(timeoutMs) {
```

Comment:
is it possible for one of the internal tasks to reach a checkpoint threshold during this period? i.e. should we wait to cancel the internal scope until after the `withTimeout` returns, so that we can still flush checkpoints

Reply:
Yes, the main reason for this is to let any persistable work in flight finish.

Keeping the internal jobs running a bit longer would probably be better for that -- the main thing would be pushing some more checkpoints through

Even more defensively, I could add another checkpoint flush after this, just to be sure every bit of goody has been got :)

Reply:
I specifically added one in the `FailSyncTask` before marking the sync failed